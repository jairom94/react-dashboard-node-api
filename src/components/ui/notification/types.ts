export type Icons = 'success' | 'info' | 'warning' | 'error'

export interface Notification {
    id:string;
    message :string;
    type: Icons;
    createdAt: number;
}

export type PrevNotification = Omit<Notification, "id" | "createdAt">