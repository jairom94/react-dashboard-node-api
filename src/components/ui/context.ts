import { createContext, useContext } from "react";

export const RadioGroupContext = createContext<{
    sale?:boolean;
    value?:string;
    name:string;
    onChange:(nameSelected:string)=>void;
}>({
    // sale:false,
    name:'',
    value:'',
    onChange:()=>{},
})

export function useRadioGroup() {
    const radioGroupValue = useContext(RadioGroupContext)
    return radioGroupValue
}

export const CheckGroupContext = createContext<{
    name:string;
    selectsItem:string[];
    onSelect:(value:string)=>void;
    onUnSelect:(value:string)=>void;
    clearSelects:()=>void;    
}>({
    name:'',
    selectsItem:[],
    onSelect:()=>{},
    onUnSelect:()=>{},
    clearSelects:()=>{},    
})

export function useCheckGroup() {
    const checkGroupValue = useContext(CheckGroupContext)
    return checkGroupValue
}