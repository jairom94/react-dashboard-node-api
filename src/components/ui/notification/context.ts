import { createContext, useContext } from "react";
import type { Notification } from "./types";

interface NotificationContextType{
    addNoti:(notification:Notification)=>void;
}

export const NotificationContext = createContext<NotificationContextType | null>(null)

export function useNotification(){
    const context = useContext(NotificationContext)
    if(!context){
        throw new Error('useNotification must be used within NotificationProvider')        
    }
    return context
}