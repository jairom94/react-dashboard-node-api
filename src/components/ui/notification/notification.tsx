import { useEffect, useState } from "react";
import type { Icons } from "./types";


interface NotificationProps {
    message: string; 
    type: Icons; 
    onDelete:()=>void; 
    createdAt:number;  
}
const Notification = ({ message, type, onDelete,createdAt }: NotificationProps) => { 
  const [progress, setProgress] = useState(0);
  const duration = 5000
  useEffect(() => {
    const intervalMs = 10;

    const interval = setInterval(() => {
      const elapsed = Date.now() - createdAt;
      const percentage = Math.min((elapsed / duration) * 100,100);
      setProgress(percentage);

      if (elapsed >= duration) {
        clearInterval(interval);
        onDelete();
      }
    }, intervalMs);

    return () => clearInterval(interval);
  }, [createdAt, onDelete]); 
  
  const icons = {
  success: "M8.5 11.5 11 14l4-4m6 2a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z",
  info: "M10 11h2v5m-2 0h4m-2.592-8.5h.01M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z",
  warning: "M12 13V8m0 8h.01M21 12a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z",
  error: "m15 9-6 6m0-6 6 6m6-3a9 9 0 1 1-18 0 9 9 0 0 1 18 0Z",
}; 
  return (
    <li className={`notification-item ${type}`}>        
      <div className="notification-content">
        <div className="notification-icon">
          <svg aria-hidden="true" fill="none" viewBox="0 0 24 24">
            <path
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              d={icons[type]}
            />
          </svg>
        </div>
        <div className="notification-text">{message}</div>
      </div>
      <button 
      onClick={onDelete}
      className="notification-icon notification-close">
        <svg aria-hidden="true" fill="none" viewBox="0 0 24 24">
          <path
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            d="M6 18 17.94 6M18 18 6.06 6"
          ></path>
        </svg>
      </button>
      <div
        className="notification-progress-bar"
        style={{ width: `${progress}%` }}
      />
    </li>
  );
};

export default Notification;
