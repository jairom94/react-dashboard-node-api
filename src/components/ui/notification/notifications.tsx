import { createPortal } from 'react-dom';
import NotificationComponent from './notification';
import type { Notification } from './types';


interface NotficationsProps {
    notifications:Notification[];
    handleNotifications:(notifications:Notification[])=>void;
}
const Notfications = ({notifications,handleNotifications}:NotficationsProps) => {
    const portalNotifications = document.querySelector('#notifications-root')
    function handleClickDelete(notificationId:string){
        const deleteNotis = notifications.filter(n => n.id !== notificationId)
        handleNotifications(deleteNotis)
    }
    return createPortal(
    <ul className='notification-container'>
        { notifications.map(notification => (
            <NotificationComponent 
            key={notification.id} 
            onDelete={()=>handleClickDelete(notification.id)} 
            message={notification.message} 
            createdAt={notification.createdAt}
            type={notification.type} 
            />
        )) }
    </ul>
    ,portalNotifications as HTMLDivElement)
};

export default Notfications;
