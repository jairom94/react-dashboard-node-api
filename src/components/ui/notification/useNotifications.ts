import { useCallback, useRef, useState } from "react";
import type { Notification } from "./types";


function useNotifications(){
    const [notifications,setNotifications] = useState<Notification[]>([])
      const timeoutMapRef = useRef<Map<string,number>>(new Map());

    const removeNoti = useCallback((notificationId:string)=>{
        setNotifications(prev => prev.filter(n => n.id !== notificationId));
        const timeoutId = timeoutMapRef.current.get(notificationId)
        if(timeoutId){
            clearTimeout(timeoutId)
            timeoutMapRef.current.delete(notificationId)
        }
    },[])

    const addNoti = useCallback((newNoti:Notification) => {
        setNotifications([...notifications,newNoti])

        const timeoutId = setTimeout(() => {
            removeNoti(newNoti.id)
        }, 5000);

        timeoutMapRef.current.set(newNoti.id,timeoutId)
    },[removeNoti,notifications])
    

    const handleNotifications = useCallback((currentNotifications: Notification[]) => {
        setNotifications(currentNotifications);
    }, []);

    return { notifications,addNoti,handleNotifications }
}

export default useNotifications