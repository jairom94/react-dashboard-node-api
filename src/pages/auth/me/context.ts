import { createContext, useContext } from "react";
import type { UserResponse } from "../types";

export const UserContext = createContext<{
    user:UserResponse | null;
    onUserLogged:()=>void;
    onUserLogout:()=>void;
}>({
  user:null,
  onUserLogged: ()=>{},
  onUserLogout: ()=>{},
});

export function useUserInformation() {
  const user = useContext(UserContext);
  return user;
}
