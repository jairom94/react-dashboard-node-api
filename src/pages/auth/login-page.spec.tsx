import { render } from "@testing-library/react";
import LoginPage from "./login-page";
import { Provider } from "react-redux";
import { NotificationContext } from "../../components/ui/notification/context";
import { Router } from "react-router";

describe("LoginPage", () => {
  const state = {
    auth: false,
    session: "",
    adverts: {
      loaded: false,
      data: [],
    },
    tags: [],
    filters: {
      name: "",
      tags: [],
      price: [0, 0],
      range: [0, 0],
    },
    ui: {
      pending: false,
      error: null,
    },
    modal: {
      data: null,
      type: "",
      visible: false,
    },
  };
  const renderComponent = () =>
    render(
      <Provider
        store={{
          getState: () => state,
          // @ts-expect-error: we only need to mock the methods we use
          subscribe: () => {},
          // @ts-expect-error: we only need to mock the methods we use
          dispatch: () => {},
        }}
      >
        <NotificationContext.Provider value={{ addNoti: vi.fn() }}>
            <Router navigator={{createHref:()=>'',go:()=>{},push:()=>{},replace:()=>{}}} 
            location={{ pathname: "" }}>
                <LoginPage />
            </Router>
        </NotificationContext.Provider>
      </Provider>,
    );
  test("should render correctly", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("")
});
