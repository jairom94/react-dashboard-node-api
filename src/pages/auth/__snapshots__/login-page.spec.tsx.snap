// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginPage > should render correctly 1`] = `
<div>
  <div
    class="min-h-dvh flex justify-center items-center bg-gray-800"
  >
    <div
      class="flex flex-col gap-3 rounded-lg bg-white px-6 py-8"
    >
      <h3
        class="text-2xl font-medium text-gray-800"
      >
        Formulario de inicio de sesión
      </h3>
      <form
        class="flex flex-col gap-2"
      >
        <div
          class="form-field flex flex-col gap-1"
        >
          <label
            for="email"
          >
            E-mail
          </label>
          <input
            class="rounded-md border px-4 py-2 undefined"
            id="email"
            name="email"
            type="email"
            value=""
          />
        </div>
        <div
          class="form-field flex flex-col gap-1"
        >
          <label
            for="password"
          >
            Contraseña
          </label>
          <input
            class="rounded-md border px-4 py-2 undefined"
            id="password"
            name="password"
            type="password"
            value=""
          />
        </div>
        <div
          class="form-field flex items-center gap-2"
        >
          <label
            for="remember"
          >
            Recordar contraseña
          </label>
          <input
            class="undefined"
            id="remember"
            name="remember"
            type="checkbox"
          />
        </div>
        <div
          class="form-field flex flex-col"
        >
          <input
            class="rounded-md border px-4 py-2 bg-emerald-600 hover:bg-emerald-500 transition-colors duration-300 cursor-pointer disabled:opacity-50 disabled:pointer-events-none"
            name="enviar"
            type="submit"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
